### PR TITLE
Employee-User attachment

### PR DESCRIPTION
**Pull request by andriyshevchenko**
Add a couple of methods to EmployeeService and EmployeeController
Small Edit on many files
Wrote unit tests

### CODE REVIEW

File: TransIT.API/Controllers/EmployeeController.cs [hunk 1]
```diff
@@ -110,6 +110,64 @@ public async Task<IActionResult> Create([FromBody] EmployeeDTO employeeDTO)
...
+            var deleteResult = await _employeeService.RemoveUserAsync(employee);
+            if (deleteResult != EmployeeDTO.DoesNotExist)
+            {
+                return NoContent();
+            }
+            else
+            {
+                return BadRequest("Employee doesn't exist");
+            }
+        }
```
Comment (igolovchuk) [inline TransIT.API/Controllers/EmployeeController.cs:218]: add try catch for all endpoints, return 500 in case of exception, log the exception using logger

Comment (andriyshevchenko) [reply TransIT.API/Controllers/EmployeeController.cs:218]: @igolovchuk Can I use exception filter? It is already in the project in EndpointFilters folder

File: TransIT.BLL/DTOs/EmployeeDTO.cs [hunk 2]
```diff
@@ -2,12 +2,23 @@
 {
     public class EmployeeDTO
     {
+        /// <summary>
+        /// Using a null object pattern instead of null.
```
Comment (igolovchuk) [inline TransIT.BLL/DTOs/EmployeeDTO.cs]: Do you use these static properties on client? If no, remove them

File: TransIT.API/Controllers/EmployeeController.cs (hunk 1 above)
Comment (igolovchuk) [reply TransIT.API/Controllers/EmployeeController.cs:218]: yes. but check if it meets all requirements described earlier in comment

File: TransIT.BLL/Services/Interfaces/IEmployeeService.cs [hunk 3]
```diff
@@ -7,5 +9,46 @@ namespace TransIT.BLL.Services.Interfaces
...
+        /// </summary>
+        /// <param name="employee">An employee ID.</param>
+        /// <returns>A data tranfer object, without attached user.</returns>
+        Task<EmployeeDTO> RemoveUserAsync(int employee);
+
+        /// <summary>
+        /// Gets all board numbers at once.
+        /// </summary>
+        /// <returns>A list of integers.</returns>
+        Task<List<int>> GetBoardNumbers();
```
Comment (igolovchuk) [inline TransIT.BLL/Services/Interfaces/IEmployeeService.cs]: add suffix 'Async' to all methods that returns Task

File: TransIT.DAL/Repositories/BaseRepository.cs [hunk 4]
```diff
@@ -17,53 +17,54 @@ public abstract class BaseRepository<TEntity> : IBaseRepository<TEntity>, IQuery
...
         public BaseRepository(TransITDBContext context)
         {
+            _entities = context.Set<TEntity>();
             _context = context;
         }
 
         public virtual Task<TEntity> GetByIdAsync(int id)
         {
-            return ComplexEntities.SingleOrDefaultAsync(t => t.Id.Equals(id));
+            return ComplexEntities.Where(t => t.Id == id).FirstOrDefaultAsync();
```
Comment (igolovchuk) [inline TransIT.DAL/Repositories/BaseRepository.cs]: use FindAsync instead where and firstordefault or revert back

File: TransIT.DAL/Repositories/BaseRepository.cs [hunk 5]
```diff
@@ -17,53 +17,54 @@ public abstract class BaseRepository<TEntity> : IBaseRepository<TEntity>, IQuery
...
         }
 
-        public virtual Task<IEnumerable<TEntity>> GetAllAsync()
+        public async virtual Task<IEnumerable<TEntity>> GetAllAsync()
```
Comment (igolovchuk) [inline TransIT.DAL/Repositories/BaseRepository.cs]: remove async await keywords from all methods that dont need it(one-line methods)

File: TransIT.Tests/Repository/EmployeeRepositoryTests.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+using Xunit;
+
+namespace TransIT.Tests
+{
+    public class EmployeeRepositoryTests
+    {
+        [Fact]
+        public async Task Employee_Repository_Should_Add_Employee()
+        {
+            var context = CreateDbContext();
```
Comment (igolovchuk) [inline TransIT.Tests/Repository/EmployeeRepositoryTests.cs:18]: add Arrange, Act Asser comments to divide parts of test method(for all test methods)

File: TransIT.Tests/ServiceTests/EmployeeServiceTests.cs [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+using TransIT.DAL.Models.Entities;
+using TransIT.DAL.Repositories.ImplementedRepositories;
+using TransIT.DAL.UnitOfWork;
+using Xunit;
+
+namespace TransIT.Tests
+{
+    public partial class EmployeeServiceTests : IClassFixture<MapperFixture>
+    {
+        private readonly string SomeUserId = Guid.NewGuid().ToString();
```
Comment (igolovchuk) [inline TransIT.Tests/ServiceTests/EmployeeServiceTests.cs:21]: add AAA comments(desribed in previous comment)

**Review by igolovchuk — CHANGES_REQUESTED** [review]
read comments in files, 
+General: add comments to your code

File: TransIT.DAL/Repositories/BaseRepository.cs [hunk 8]
```diff
@@ -17,53 +17,54 @@ public abstract class BaseRepository<TEntity> : IBaseRepository<TEntity>, IQuery
...
         public virtual Task<TEntity> GetByIdAsync(int id)
         {
-            return ComplexEntities.SingleOrDefaultAsync(t => t.Id.Equals(id));
+            return ComplexEntities.SingleOrDefaultAsync(entity => entity.Id == id);
         }
 
         public virtual Task<IEnumerable<TEntity>> GetAllAsync()
         {
-            return Task.FromResult<IEnumerable<TEntity>>(ComplexEntities);
+            return Task.FromResult(ComplexEntities.AsEnumerable());
```
Comment (igolovchuk) [inline TransIT.DAL/Repositories/BaseRepository.cs]: why did you remove tolistasync?

File: TransIT.DAL/Repositories/BaseRepository.cs [hunk 9]
```diff
@@ -17,53 +17,54 @@ public abstract class BaseRepository<TEntity> : IBaseRepository<TEntity>, IQuery
...
         }
 
-        public virtual Task<IEnumerable<TEntity>> GetAllAsync(Expression<Func<TEntity, bool>> predicate)
+        public async virtual Task<IEnumerable<TEntity>> GetAllAsync(Expression<Func<TEntity, bool>> predicate)
```
Comment (igolovchuk) [inline TransIT.DAL/Repositories/BaseRepository.cs]: remove async await keywords, just keywords

File: TransIT.DAL/Repositories/BaseRepository.cs [hunk 10]
```diff
@@ -17,53 +17,54 @@ public abstract class BaseRepository<TEntity> : IBaseRepository<TEntity>, IQuery
...
         }
 
         public virtual TEntity Update(TEntity entity)
         {
-            return Entities.Update(entity).Entity;
+            return _entities.Update(entity).Entity;
         }
 
-        public virtual Task<IEnumerable<TEntity>> GetRangeAsync(uint index, uint amount)
+        public async virtual Task<IEnumerable<TEntity>> GetRangeAsync(uint index, uint amount)
```
Comment (igolovchuk) [inline TransIT.DAL/Repositories/BaseRepository.cs]: emove async await keywords, just keywords

File: TransIT.API/Controllers/EmployeeController.cs [hunk 11]
```diff
@@ -106,7 +112,76 @@ public async Task<IActionResult> Create([FromBody] EmployeeDTO employeeDTO)
...
             {
-                throw e;
+                _logger.LogError(e, string.Empty);
+                return StatusCode(500, new ExtendedErrorDTO(e));
+            }
+        }
+
+        [HttpGet("boardnumber/{number}")]
+        public async Task<IActionResult> GetByBoardNumber(int number)
+        {
```
Comment (igolovchuk) [inline TransIT.API/Controllers/EmployeeController.cs:122]: I don't see changes: add try catch for **ALL** endpoints you've added, return 500 in case of exception, log the exception using logger

File: TransIT.DAL/Repositories/BaseRepository.cs (hunk 9 above)
Comment (andriyshevchenko) [reply TransIT.DAL/Repositories/BaseRepository.cs]: Because , when i remove await, it cannot convert Task<List<item> to task Task<IEnumerable<item>>

File: TransIT.DAL/Repositories/BaseRepository.cs (hunk 8 above)
Comment (andriyshevchenko) [reply TransIT.DAL/Repositories/BaseRepository.cs]: @igolovchuk Because i cannot convert Task<List<Employee>> to Task<IEnumerable<Employee>>

File: TransIT.API/Controllers/EmployeeController.cs [hunk 12]
```diff
@@ -41,7 +44,8 @@ public async Task<IActionResult> Get([FromQuery] uint offset = 0, uint amount =
             }
             catch (Exception e)
             {
-                return StatusCode(500, e);
+                _logger.LogError(e, string.Empty);
```
Comment (igolovchuk) [inline TransIT.API/Controllers/EmployeeController.cs]: please do not log string.empty, log e.Messege

Comment (andriyshevchenko) [reply TransIT.API/Controllers/EmployeeController.cs]: Changed.

**Review by igolovchuk — APPROVED** [review]

**Review by Vladacdc — APPROVED** [review]

**Review by VitalikRibiy — APPROVED** [review]
